### DISCUSSION TITLE
Detecting timestamp for the start of speech within a file

### DISCUSSION THREAD

**Discussion by gtebbutt** [openai/whisper#237] [Q&A] [answered]
We've been having some trouble syncing the output timestamps to the start of speech if it doesn't precisely coincide with the beginning of the file, or after a pause within the file, and I was wondering if anyone had insight into why this would be or ideas for how to work around it? Overall results have been very impressive, both accuracy and sync within speech, it's just starting points that are proving to be an issue. 

As a hypothetical example: take a file that begins with ten seconds of music, followed by the words "Four score and seven years ago" from timestamp `00:00:10` to `00:00:14`. The output we're seeing would look like `[00:00:00 --> 00:00:14] Four score and seven years ago` - but we actually want that first timestamp to be `00:00:10` to avoid the subtitle hanging on screen for ten seconds before anyone begins talking. We're seeing the same after pauses that happen within the file, but the start and end times for a given line within continuous speech are spot on. 

Looking into the token-level timestamps with https://github.com/jianfch/stable-ts is interesting (big thank you to @jianfch!) - we'll see something like the following: 
```
[00:00:00.500] Four
[00:00:06.000] score
[00:00:10.300] and
[00:00:10.870] seven
...
```

So the first couple of words are individually being timestamped before the start of speech, with gaps far too large, but then the later words come more or less into sync, and by the end of the line or phrase the timestamp of the last word is pretty much dead on. 

A potential workaround is just removing any non-speech audio with something like Silero VAD (as per #29) and stitching the files back together afterwards, but I'd still like to try and understand the issue first. Any insight would be very much appreciated!

**Comment by jianfch** [comment, marked as answer]
the script uses the start and end of the segments as min and max to select the token-level timestamps. `unstable_word_timestamps`  for that segment might reveal more as to what's going on. 'Four' and 'score' might even have timestamps that start after 10seconds.

**Reply by gtebbutt** [reply]
Thanks! I've taken a look at that and it definitely provides some starting points to explore.

The real test file has the actual speech starting at `00:00:16`. The first word's unstable timestamps are consistently inaccurate (range from `0.0` to `0.799`, so all off by at least 15 seconds), second word has a much higher variance (range from `0.0` to `19.56`, with the closest still of by two seconds) and then by the third word it starts to stabilise on a good value (range from `17.96` to `18.20`, very close to accurate). 

I've spot checked some later sections that *are* giving accurate sentence-level output, and some of those also have high variance in the unstable timestamps (over 10 seconds for a single token), so I don't think that alone will work for detecting pauses, but it's a signal I can potentially look at in combination with others.

**Reply by jongwook** [reply]
@gtebbutt The default option has a heuristically set value of `max_initial_timestamp=1.0`, which is might be related to why you're getting timestamps between 0.0 and 0.799. Setting that to `None` will remove the constraint, but the model has a huge bias to the 0.0 timestamp (and other integers) so it may not behave any better..

**Reply by gtebbutt** [reply]
Got it, thanks, that's really useful to know.

**Reply by dillfrescott** [reply]
I tried everything and the subtitles are starting at 00:00.00 when the first utterance isn't until nearly 24 seconds into the file. It's not just the beginning either. It happens many many times throughout the audio, the subtitle will appear sometimes 20 seconds or more before the beginning of said utterance.

**Reply by gtebbutt** [reply]
Yeah, we ended up going down the Silero VAD route in the end. Changing parameters within Whisper didn't have a strong impact on this specific issue, so it's easier to just preprocess the file to detect speech, slice up the audio to remove non-speech portions, run each of those through Whisper separately, and then stitch the text back together at the end.

**Reply by a-ruban** [reply]
@gtebbutt i tried to improve Whisper timecodes using Aeneas (forced alignment). It works, not very stable, but for some audios absolutely brilliantly.
Could you please share your solution with Silero VAD?
And I will prepare and share my with Aeneas so we will be able to compare the results

**Reply by arnavmehta7** [reply]
@a-ruban Hi, I personally found Aeneas very unstable. when there is some noise in audio. It completely messes up following texts after noise. How you fixed that?

**Reply by a-ruban** [reply]
@a-ruban Yeah, that's true for me also.
I compare Whisper timecodes (separately for start_time and end_time) with Aenas timecodes and if the difference in second is less than 5-6 seconds - I take Aenas timecode, if not - take Whisper timecodes.
It gives more precise timecodes in most of the cases.

**Reply by arnavmehta7** [reply]
thats not a good one @a-ruban

**Reply by a-ruban** [reply]
@arnavmehta7 why? It gives an opportunity to increase accuracy of timecodes before Aenas stumble upon noisy fragment and brake timecodes. In the worst case - you get the same timecodes that Whisper gives you, in the general case - you improve timecodes for part of the audio, in the best - improve for the whole audio.

**Reply by usmanagha125** [reply]
@a-ruban Could you please share (maybe the code) how you used Aeneas to get word-level timestamps on the plain text Whisper transcription. I've tried getting word-level timestamps using stable_whisper but the results are not very accurate. It splits some words into half. I've been trying to use Aeneas as a python library to get word-level timestamps. I am basically building an Automatic Dubbing Software as my final year project

**Reply by andupotorac** [reply]
> vel t

Use this for word-level timestamps: https://github.com/m-bain/whisperX

**Reply by luisbnzsa** [reply]
I used one feature that is on the repository in the tips section "use demucs=True to isolate vocals with [Demucs](https://github.com/facebookresearch/demucs); it is also effective at isolating vocals even if there is no music". In one of the tests I had, there was music for about 40 seconds and the timestamps at the end was 10 seconds off. After using demucs=True, it was correct.

**Reply by LaurinmyReha** [reply]
This delivered the best word level timestamps (including fillers and pauses) in our experiments. 

https://github.com/nyrahealth/CrisperWhisper
